libs/sdk-model/src/insight/sanitization.ts

```diff
@@ -1,32 +1,62 @@
 // (C) 2020 GoodData Corporation
 import { InvariantError } from "ts-invariant";
+import { uniqWith } from "lodash";
 
-import { insightSorts, IInsightDefinition, insightBuckets, insightSetBuckets } from ".";
-import { bucketTotals, IBucket, bucketAttributeIndex, bucketSetTotals } from "../execution/buckets";
-import { ISortItem, isAttributeSort, sortEntityIds, isMeasureSort } from "../execution/base/sort";
+import { IInsightDefinition, insightAttributes, insightBuckets, insightSetBuckets, insightSorts } from ".";
+import { bucketAttributeIndex, bucketSetTotals, bucketTotals, IBucket } from "../execution/buckets";
+import { isAttributeSort, isMeasureSort, ISortItem, sortEntityIds } from "../execution/base/sort";
 import { ITotal } from "../execution/base/totals";
-
+import { attributeLocalId, isAttribute } from "../execution/attribute";
+import { areObjRefsEqual } from "../objRef";
 /**
  * Makes sure the insight does not have any nonsensical data (like totals that no longer make sense, etc.), before it is saved.
  *
  * @param insight - the insight or insight definition to sanitize
  * @public
  */
 export function insightSanitize<T extends IInsightDefinition>(insight: T): T {
-    return removeInvalidTotals(insight);
+    return removeInvalidTotals(removeDuplicateDrillAttributes(insight));
+}
+
+function removeDuplicateDrillAttributes<T extends IInsightDefinition>(insight: T): T {
+    const removed = insightBuckets(insight).map((bucket) => {
+        if (bucket.localIdentifier === "attribute") {
```

Comment:
this indicates that the code is misplaced. code in sdk-model must not act on / do things based on type of bucket. if you want to do things for specific buckets, then make the functions in sdk-model accept parameters or better, let them accept function from the caller. 

then again, another valid approach is to move this particular function outside, to a better, vis-specific place. let it do partial sanitization in the context of the viz, then call the sdk-model to clean things up.

Reply:
I was now also thinking about removing this condition entirely (bucket.localIdentifier === "attribute") and just making the function generic, i.e. removeDuplicateAttributesFromBuckets() which would not be bucket.localidentifier-specific, just if it finds duplicities in a bucket, it will remove them. What would you think?

Reply:
Duplicities removal moved close to drill target replacement (https://github.com/gooddata/gooddata-ui-sdk/pull/585/commits/ab3acb2b5d3e40e8984afb41183ba4178acd6da5)